### ISSUE TITLE
Long-term: Reworked Events Page

### ISSUE THREAD

**Issue by mattxwang** [uclaacm/website#115]
Likely our most requested feature is to significantly amp-up our [Events page](https://www.uclaacm.com/events), which is currently just a Google Calendar embed. Turns out, making good events pages is ... hard.

Three limiting concerns that drive a more aggressive rework:

1. Google Calendar is a limited service: for example, we cannot embed images, and the views are not easy to customize.
2. Facebook is used less and less by incoming classes, and thus we need a way to convey event information without requiring Facebook use.
3. There's currently no central authority where general members can find all of ACM's events for a given week. Google calendar has served as a patchwork solution, but has drawbacks listed above. With a revamped events page on our website, we can offer our community this resource.

Given the large scope of this task, I want to split it up into three (mostly) disjoint things:

1. **Data**: how should we store the data for each event? do we host it internally, or use another service and hit an API? how does the membership portal play into this? How do we balance that committees will likely have different needs?
2. **Design**: how should the page look? can we make it responsive (given that a significant amount of traffic is mobile)?
3. **Interface**: can we make it easy for users to see all the relevant event information, without being overwhelmed? can we make it easy to filter for events based off of committee, type, etc.? what about people creating an event?

I want to stress that this is a **large ticket**, and it is:

* fine to be lost, since this document is **long**
* encouraged to split this up upon multiple PRs
* totally okay to ask for help!

## The Data

First, we need to decide what data is needed for each event. Examples include:

* "general" info: name of event, who's holding it, when it is
* online and/or in-person location
* blurb describing the event in detail
* relevant resources (ex GitHub, slides, recordings)
* images?

Next, we need to decide how to *store* this information. There are a few options; however, if possible, I would like to have a **single source of truth**. Here are some, off the top of my head:

* store the data in the repository. Marketing already has a spreadsheet with most relevant event information quarter-on-quarter; we can write a script to convert the CSV -> JSON, then render out of the JSON. **this is the lowest friction option**, in my opinion.
* use the **membership portal** as a single source of truth; this was the plan in #9, though I'm personally unsure of how feasible this is.
* write our own service that provides event information as an API; then, rely on this service for the membership portal AND the static website
* use Google Calendar as a "headless CMS (content management system)" for our data
* somehow use Facebook (unclear on how to do this without violating Terms of Service)
* some other approach that I haven't thought about!

The platform you choose is relatively important, since it has quite a few ripple effects on how easy it is to use the service to publish events, what information can be disseminated, costs, maintainability, etc. More than happy to discuss this more in detail, as this is a relatively complicated decision.

One thing to consider: we want to **decouple the data from the interface** as much as possible. In particular, this means that we can swap out our interface without affecting our data solution, or change our data provider without having to refactor our frontend.

Then, you need to hook up the data to the design!

## The Design 

This is something that we can definitely lean on outside help if you'd like (as referenced in #105), but I think the design of this page is critical. Some things to keep in mind:

* it should, in general, fit within our branding guidelines
* it should provide enough relevant information **without** overwhelming the user
* it should work across multiple viewports: portrait on a phone, tablets, small laptops, and large monitors
* it should be somewhat customizable, in that users should be able to choose a compact view (ex list) and a larger, comfortable view (ex week/month calendar)
* accessibility here is quite key, given this is a portal - can you use the events page with just a keyboard? 

Given the large scope of this project, I would 100% expect this to be an iterative design-develop process: start with a skeleton, slowly implement, get feedback, etc.

Event page design is not a new problem; working with your designer (or independently, if you'd like), you can scour the internet for event page examples. Keep in mind that ours are mostly workshops/learning (and not concerts, etc.) so small interface details may be different.

## The Interface

There are two interfaces we need to consider.

On the user/attendee end,

* how easy is it for the user to find the information they need? can we develop intuitive filters, search bars, etc.
* can they find out more information easily?
* what things does the user *have* to have to use our events page? Facebook? etc.
* are all interface elements relatively self-explanatory? are the defaults reasonable?
* is the interface accessible? what if fonts are disabled, the user only uses a keyboard/screenreader, is contrast good enough
* how do we deal with localization (ex timezones)?
* does the interface load quickly?

On the event creator's end,

* can someone make an event **without needing to know how to code**
* is it obvious what information is needed for an event?
* what happens if there isn't a reasonable entry for a certain field (ex location)? do we validate inputs?
* is editing events easy? Is it seamless to add or update new events every week? 
* is the interface extendable for future changes?

## Examples

Some student org examples to draw inspiration from:

* [ACM Hack's Events Page](https://hack.uclaacm.com/events)
* [ACM Cyber's Events Page](https://acmcyber.com/events)
* the [UCLA Community Events Page](https://community.ucla.edu/)

---

Supersedes #105, #9.

**Comment by mattxwang** [comment]
Backend data source is mostly blocked by membership-portal work!

**Comment by advaithg** [comment]
Currently unblocked, as implementation should be mostly decoupled from the data and we should be able to get started on this!
For now, we can just use a JSON file to generate data and ensure we have a working page - which can later be modified to work with the endpoint

**Comment by mattxwang** [comment]
Might want to consider Jason Quense's [react-calendar](https://github.com/jquense/react-big-calendar)?
